src/main/scala/org/allenai/plugins/DeployPlugin.scala

```diff
@@ -236,64 +237,170 @@ object DeployPlugin extends AutoPlugin {
+      // Strip leading / from deployTarget so we can match names on it.
+      val namePattern = deployTarget.tail
+      // Build a regex matching the replicas we will eventually update / restart as part of this
+      // process, and therefore DON'T want to stop and remove.
+      val restartedPattern = if (numReplicas == 1) {
+        namePattern
+      } else {
+        s"$namePattern-[1-$numReplicas]"
+      }
+      // Use find -exec to locate, stop, and remove stale replicas that won't be updated as part
+      // of the current deploy.
+      val findCommand = Seq(
+        "find",
+        s"$deployParent/*",
+        "-prune",
+        "-type", "d",
+        "-name", "\"" + s"$namePattern*" + "\"",
+        "-not", "-name", "\"" + restartedPattern + "\"",
+        "-exec", s"{}/bin/$namePattern.sh stop \\;",
+        "-exec", "rm -r {} \\;"
+      )
+      val stopCommand = Seq.concat(sshCommand, Seq(deployHost), findCommand)
+      // Shell-friendly version of find command, for logging.
+      val quotedStopCommand = Seq.concat(
+        sshCommand,
+        Seq(deployHost, findCommand.mkString("'", " ", "'"))
+      )
+
+      log.info("Running " + quotedStopCommand.mkString(" ") + " . . .")
+      if (Process(stopCommand).! != 0) {
+        sys.error(s"Error while trying to tear down stale replicas on '$deployHost'.")
+      }
 
-          log.info("Running " + restartCommand.mkString(" ") + " . . .")
-          if (Process(restartCommand).! != 0) {
-            sys.error(s"Error running restart command on host '$deployHost'.")
+      replicas.zipWithIndex map {
+        case (replicaConfig, i) => {
+          Future {
+            // Determine the specific remote directory for this replica.
+            val deployDirectory = if (numReplicas > 1) {
+              s"$baseDeployDir-${i + 1}"
+            } else {
+              baseDeployDir
+            }
+
+            // Command to pass to rsync's "rsync-path" flag to ensure creation of the remote dir.
+            // http://www.schwertly.com/2013/07/forcing-rsync-to-create-a-remote-path-using-rsync-path/
+            val pathCommand = Seq(
+              "mkdir",
+              "-p",
+              deployDirectory,
+              "&&",
+              "rsync"
+            )
+
+            val rsyncDirs = deployDirs.value map { name => s"--include=/$name" }
+            val rsyncCommand = Seq.concat(
+              Seq(
+                "rsync",
+                "-vcrtzP",
+                s"--rsync-path=${pathCommand.mkString(" ")}",
+                s"--rsh=${sshCommand.mkString(" ")}"
+              ),
+              rsyncDirs,
+              Seq(
+                "--exclude=/*",
+                "--delete",
+                universalStagingDir.getPath + "/",
+                deployHost + ":" + deployDirectory
+              )
+            )
+            // Shell-friendly version of rsync command, with rsh value quoted.
```

Comment:
This is confusing.

It's being quoted so you can print it out & allow copy-pasting, yes? You may wish to comment on that.

It would also be less confusing if this were created closer to use.